contracts/yearnV1/vault-savings/VaultSavings.sol

```diff
@@ -0,0 +1,182 @@
+    using SafeMath for uint256;
+
+    struct VaultInfo {
+        bool isActive;
+        uint256 blockNumber;
+    }
+
+    address[] internal registeredVaults;
+    mapping(address => VaultInfo) vaults;
+
+    address registry;
+
+    function initialize(address _registry) public onlyOwner {
+        registry = _registry;
+    }
+    
+
+
+    // deposit, withdraw
+
+    function deposit(address[] calldata _vaults, uint256[] calldata _amounts) external override nonReentrant  {
+        require(_vaults.length == _amounts.length, "Size of arrays does not match");
+
+        for (uint256 i=0; i < _vaults.length; i++) {
+            _deposit(_vaults[i], _amounts[i]);
+        }
+    }
+
+    function deposit(address _vault, uint256 _amount) external override nonReentrant returns(uint256 lpAmount) {
+        lpAmount = _deposit(_vault, _amount);
+    }
+   
+
+    function _deposit(address _vault, uint256 _amount) internal returns(uint256 lpAmount) {
+        //check vault
+        require(isVaultRegistered(_vault), "Vault is not Registered");
+
+        ( , address baseToken,  ,  ,) = IYRegistry(registry).getVaultInfo(_vault);
+     
+        //transfer token if it is allowed to contract
+        IERC20(baseToken).safeTransferFrom(msg.sender, address(this), _amount);
+
+        //set allowence to vault
+        IERC20(baseToken).safeIncreaseAllowance(_vault, _amount);
+
+        //deposit token to vault
+        IVault(_vault).deposit(_amount);
+
+        lpAmount = IERC20(_vault).balanceOf(address(this));
+        //send new tokens to user
+        IERC20(_vault).safeTransfer(msg.sender, lpAmount);
+
+        emit  Deposit(_vault, msg.sender, _amount, lpAmount);
+    }
+
+
+    function withdraw(address[] calldata _vaults, uint256[] calldata _amounts) external override nonReentrant {
+        require(_vaults.length == _amounts.length, "Size of arrays does not match");
+
+        for (uint256 i=0; i < _vaults.length; i++) {
+            _withdraw(_vaults[i], _amounts[i]);
+        }
+
+    }
+
+    function withdraw(address _vault, uint256 _amount) external override nonReentrant returns(uint256 baseAmount) {
+        baseAmount = _withdraw(_vault, _amount);
+    }
+
+    function _withdraw(address _vault, uint256 _amount) internal returns(uint256 baseAmount) {
+        require(isVaultRegistered(_vault), "Vault is not Registered");
+        //transfer LP Token if it is allowed to contract
+        IERC20(_vault).safeTransferFrom(msg.sender, address(this), _amount);
+
+        //burn tokens from vault
+        IVault(_vault).withdraw(_amount);
+
+        ( , address baseToken,  ,  ,) = IYRegistry(registry).getVaultInfo(_vault);
+
+        baseAmount = IERC20(baseToken).balanceOf(address(this));
+
+        //Transfer token to user
+        IERC20(baseToken).safeTransfer(msg.sender, baseAmount);
+
+        emit Withdraw(_vault, msg.sender, baseAmount, _amount);
+    }
+
+    function registerVault(address _vault) external override {
```

Comment:
Maybe we can restrict who can call this to `onlyOwner`?

Reply:
Agree